chaincode/compute_plan.go

```diff
@@ -319,5 +320,104 @@ func getComputePlan(db *LedgerDB, key string) (resp outputComputePlan, err error
 		AggregatetupleKeys:      aggregatetupleKeys,
 		TesttupleKeys:           testtupleKeys,
 	}
+
+	resp.Status, err = getComputePlanStatus(db, resp)
+	if err != nil {
+		return
+	}
+
 	return
 }
+
+func getComputePlanStatusByComputePlanID(db *LedgerDB, computePlanID string) (status string, err error) {
+	computePlan, err := getComputePlan(db, computePlanID)
+	if err != nil {
+		return "", err
+	}
+
+	return getComputePlanStatus(db, computePlan)
+}
+
+func getComputePlanStatus(db *LedgerDB, computePlan outputComputePlan) (status string, err error) {
+	// get all tuples like status in one slice
+	statusCollection := []string{}
+
+	keys := []string{}
+	keys = append(keys, computePlan.TraintupleKeys...)
+	keys = append(keys, computePlan.CompositeTraintupleKeys...)
+	keys = append(keys, computePlan.AggregatetupleKeys...)
+	keys = append(keys, computePlan.TesttupleKeys...)
+
+	for _, key := range keys {
+		tuple, err := db.GetGenericTuple(key)
+		if err != nil {
+			return "", err
+		}
+
+		statusCollection = append(statusCollection, tuple.Status)
+	}
+
+	return determineComputePlanStatus(statusCollection)
+}
+
+func determineComputePlanStatus(statusCollection []string) (status string, err error) {
+	// this status order matters
+	sts := []string{StatusCanceled, StatusFailed, StatusDoing, StatusTodo, StatusWaiting, StatusDone}
+	for _, s := range sts {
+		if stringInSlice(s, statusCollection) {
+			return s, nil
+		}
+	}
+
+	return "", fmt.Errorf("unknown compute plan status")
+}
+
+func cancelComputePlan(db *LedgerDB, args []string) (resp bool, err error) {
+	computeplan, err := queryComputePlan(db, args)
+	if err != nil {
+		return
+	}
+
+	tupleKeys, err := db.GetIndexKeys("computePlan~computeplanid~worker~rank~key", []string{"computePlan", computeplan.ComputePlanID})
+	if err != nil {
+		return
+	}
+	if len(tupleKeys) == 0 {
+		err = errors.E("no task found for compute plan %s", computeplan.ComputePlanID)
+		return
+	}
+
+	for _, key := range tupleKeys {
+		tuple, err := db.GetGenericTuple(key)
+		if err != nil {
+			return false, err
+		}
+		err = tuple.commitStatusUpdate(db, key, StatusCanceled)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	testtupleKeys := []string{}
+	for _, key := range tupleKeys {
+		keys, err := db.GetIndexKeys("testtuple~traintuple~certified~key", []string{"testtuple", key})
+		if err != nil {
+			return false, err
+		}
+
+		testtupleKeys = append(testtupleKeys, keys...)
+	}
+
+	for _, key := range testtupleKeys {
+		testtuple, err := db.GetTesttuple(key)
+		if err != nil {
+			return false, err
+		}
+		err = testtuple.commitStatusUpdate(db, key, StatusCanceled)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	return true, nil
```

Comment:
Like I said directly to @inalgnu, the return value of a smartcontract should respect the [interface `Unmarshaler`](https://golang.org/pkg/encoding/json/#Unmarshaler).

Reply:
done, thank's